### ISSUE TITLE
functools.partial is not supported(?)

### ISSUE THREAD

**Issue by gnarlyquack** [facebook/pyre-check#199]
I seem to have come across a couple issues with using functools.partial:

```python
# pyre-strict
import functools
import typing


def callback(name: str) -> None:
    print('Hello', name)


def do_callback(fun: typing.Callable[[], None]) -> None:
    fun()


def main() -> None:
    f1 = functools.partial(callback, 5)
    do_callback(f1)

    f2 = functools.partial(callback)
    do_callback(f2)

```

I would expect incompatible parameter type errors to be raised both for the definition of f1 and when attempting to invoke do_callback with f2. As of pyre version 0.0.30, no type errors are raised.

**Comment by pradeep90** [comment]
This is because [`typeshed`](https://github.com/python/typeshed/blob/ec7960a8cb2e2ff465ee0fff841cd28fa85fe3ba/stdlib/2/functools.pyi#L28) types the callable as expecting arbitrary parameters: 

```
class partial(Generic[_T]):
    func = ...  # Callable[..., _T]
    args: Tuple[Any, ...]
    keywords: Dict[str, Any]
    def __init__(self, func: Callable[..., _T], *args: Any, **kwargs: Any) -> None: ...
```

You can look into Pyre's [`pyre_extensions`](https://pypi.org/project/pyre-extensions/) for `ParameterSpecification` to type the above code.
